buzzmobile/process/bearing/bearing.py

```diff
@@ -25,31 +25,71 @@ def median(self):
             return (sortedArray[index] + sortedArray[index + 1]) / 2.0
 
 
-#Global Variables
-bearings = {}
-bearings['last_fix'] = None
-bearings['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
+# Global variables
+globalVars = {}
+globalVars['last_fix'] = None
+globalVars['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
 bearing_pub = rospy.Publisher('bearing', Float64, queue_size=1)
 
+MIN_FIX_DISTANCE = rospy.get_param('min_fix_distance')
+EARTH_RADIUS = 6.3710088e6
+
+
 def bearing(fix):
-    global last_fix
-    global med_filter
-
-    if bearings['last_fix'] is not None:
-        lat1 = math.radians(bearings['last_fix'].latitude)
-        lon1 = math.radians(bearings['last_fix'].longitude)
-        lat2 = math.radians(fix.latitude)
-        lon2 = math.radians(fix.longitude)
-        y = math.sin(lon2 - lon1) * math.cos(lat2)
-        x = math.cos(lat1)*math.sin(lat2) - math.sin(lat1)*math.cos(lat2)*math.cos(lon2 - lon1)
-        angle = math.atan2(y, x)
-        computedBearing = (angle + 2*math.pi) % 2*math.pi
-        bearings['med_filter'].add(computedBearing)
-
-        bearing_pub.publish(bearings['med_filter'].median())
+    """
+    Adds the latest computed bearing to the meadian filter, publishes
+    the current median bearing, and sets last_fix to the given fix
+    if the given fix is >= MIN_FIX_DISTANCE from last_fix or there is
+    currently no last_fix
+    """
+
+    last_fix = globalVars['last_fix']
+    med_filter = globalVars['med_filter']
 
     if fix is not None:
-        last_fix = fix
+        if last_fix is not None:
+            distance = get_distance(last_fix, fix)
+            bearing = get_forward_angle(last_fix, fix)
+            med_filter.add(bearing)
+            bearing_pub.publish(med_filter.median())
+            if distance >= MIN_FIX_DISTANCE:
+                last_fix = fix
+        else:
+            last_fix = fix
+    globalVars['last_fix'] = last_fix
```

Comment:
Because it doesn't get modified. The problem with global variables and ros is that if you have the following:

```python
a_global

def message_cb(new_msg):
    a_global = new_msg
```

What's actually happening is that the `a_global` inside `message_cb` is not the global variable `a_global`, but a local variable to the method `message_cb`. This is not a problem with ros publishers because:

```python
pub = rospy.Publisher('msg_name', String, queue_size=1)

def some_method():
    pub.publish('hey')
```

`pub.publish()` would fails if `pub` is a local variable, so python looks up if there's a global named `pub` and finds the publisher we defined.

I don't know why the first problem happens exactly, but it has to do with ros+python interactions because it doesn't happen in vanilla python.